app/scripts/background.js

```diff
@@ -981,45 +981,88 @@ export function setupController(
           connectionStream: portStreamForCookieHandlerPage,
         });
       }
-      connectExternalExtension(remotePort);
+
+      const portStream =
+        overrides?.getPortStream?.(remotePort) || new PortStream(remotePort);
+
+      connectEip1193(portStream, remotePort.sender);
+
+      if (process.env.MULTICHAIN_API && isFirefox) {
+        const mux = setupMultiplex(portStream);
+        mux.ignoreStream(METAMASK_EIP_1193_PROVIDER);
+
+        connectCaipMultichain(
+          mux.createStream(
+            METAMASK_CAIP_MULTICHAIN_PROVIDER,
+            remotePort.sender,
+          ),
+        );
+      }
     }
   };
 
-  // communication with page or other extension
-  connectExternalExtension = (remotePort) => {
+  /**
+   * Connects a externally_connecatable Port to the MetaMask controller.
+   * This method identifies dapp clients and connects them differently from extension clients.
+   *
+   * @param {Port} remotePort - The port provided by a new context.
+   */
+  connectExternallyConnectable = (remotePort) => {
     const portStream =
       overrides?.getPortStream?.(remotePort) || new PortStream(remotePort);
+
+    const isDappConnecting = remotePort.sender.tab?.id;
+    if (isDappConnecting && process.env.MULTICHAIN_API) {
+      if (metamaskBlockedPorts.includes(remotePort.name)) {
+        return;
+      }
+
+      // this is triggered when a new tab is opened, or origin(url) is changed
+      if (remotePort.sender && remotePort.sender.tab && remotePort.sender.url) {
+        trackDappView(remotePort);
+      }
+
+      connectCaipMultichain(portStream, remotePort.sender);
+    } else {
+      connectEip1193(portStream, remotePort.sender);
+    }
+  };
+
+  /**
+   * Connects a Duplexstream to the MetaMask controller EIP-1193 API (via a multiplexed duplex stream).
+   *
+   * @param {DuplexStream} connectionStream - The duplex stream.
+   * @param sender - The remote port sender.
```

Comment:
I _think_ sender is optional? Not 100% sure though. I believe the type is:

```suggestion
   * @param {chrome.runtime.MessageSender} [sender] - The remote port sender.
```

Reply:
i'll check

Reply:
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/Port#sender

indeed it is optional, but defined when coming from `browser.runtime.onConnect` and `browser.runtime.onConnectExternal`. So it will be defined for our use cases

Reply:
Can you add the correct type for `sender` here (and below).

Reply:
oops. yes https://github.com/MetaMask/metamask-extension/pull/30142/commits/4e1c5e1c7fa6447d30828603bdf102dbfa44748c